Migrate Release pipeline to GitHub Actions/Xcode Cloud

dianaafanador3 opened the pull request:
Added two new Github Actions, which replace the two Jenkins pipelines which were used for release.

The release workflows now will happen in two steps
1. When a push to master is made (Github action master-push.yml)
This GitHub actions creates the release packages, including docs, and store it in the artifacts repository in GitHub Actions for later use.

Previously packaging happened in a two step process, which consisted in a first step consisting in building for each platform, target and configuration (static for iOS) and create the xcframework for each platform and xcode version, and a second step where using the created xcframeworks in the previous step, we create the final release package. 

The new approach is a 3 step process, and just divides the first step into two step from the previous approach.
- Creates a framework for each combination of supported (platform, xcode version, target and configuration). This will use Xcode Cloud but I also added a step (commented at the moment) which will enable this with GitHub actions.
- Combine every framework generated for each platform (platform and simulator if needed), target (Realm, RealmSwift) and configuration (Release and static for iOS) by Xcode version and generate the xcframework.
- Combine all the platform xcframework and versions and generating the final packages.
We are doing this in-paralell which probably will make the release times faster, depending on Xcode Cloud times.

Installation test are running in the GitHub Actions runner for easier run. This tie us to use an older version of the Xcode sometimes, because the runners are not updated as quick as in Xcode Cloud, but the reason is that we need the release package for some of this test (specifically xcframework test) which will need to set an environment value into the Xcode Cloud workflow and will make configuring the release workflow harder.
GitHub Actions runners are updated after a few weeks max after the release of the Xcode version version (not betas). 

2. Publishing a release is executed in a second step, this workflow has to be executed manually from GitHub Actions (Github action publish-release.yml)
This workflow executes all steps for release, tagging, create the GitHub release, publish to cocoa pods, update the file checker, publish docs, after that it will run some installation test over the release and finally it will post the release to the release slack channel.

Note: I'm reverting some changes I did for the PR pipelines migration, this changes were affecting some test in the release pipeline and shouldn't have been done, the only thing needed was to add the Static configuration for each of the targets in the RealmExamples project.

Xcode Cloud usage is at its limits so probably this cannot be merged until after the 28.

tgoyne reviewed (commented):
A first glance over things.

dianaafanador3 commented on .github/workflows/master-push.yml:
```diff
@@ -0,0 +1,227 @@
+                # Key ID. Can get from App Store Connect.
+                key id: ${{ secrets.APPLE_STORE_CONNECT_KEY_ID }}
+                # P8 private key. Can get from App Store Connect.
+                key:  ${{ secrets.APPLE_STORE_CONNECT_API_KEY }}
```
Apple recommend to generate tokens instead of creating a long lived one
https://developer.apple.com/documentation/appstoreconnectapi/generating_tokens_for_api_requests#3878467

tgoyne replied:
With the changes in #8438 I guess this is irrelevant as we can just pass the secrets directly to the script and skip the separate Get Token step.

dianaafanador3 replied:
mmm the script arguments accepts a private key path, we will have to add something to be able to pass the private key as value like we have it stored in GitHub secret, instead of a path in that case

dianaafanador3 replied:
but I can add that

dianaafanador3 commented on .github/workflows/master-push.yml:
```diff
@@ -0,0 +1,227 @@
+jobs:
+  prepare:
+      runs-on: ubuntu-latest
+      name: Create Draft Release to store artifacts
```
I'm transitioning to cache for inside a workflow storage and GitHub Actions artefact storage for publish artefacts. The APIs for this are private and the easiest way to access it is through a template action.

dianaafanador3 commented:
@tgoyne changed the approach for the release, updated the PR's descriptions with the new approach

tgoyne commented on .github/workflows/master-push.yml:
```diff
@@ -0,0 +1,564 @@
+            gh extension install actions/gh-actions-cache
+            cacheKeysForPR=$(gh actions-cache list -R $REPO -B $BRANCH -L 100 | cut -f 1 )
+                  
+            ## Setting this to not fail the workflow while deleting cache keys. 
+            set +e
```
`|| true` on the specific things which are allowed to fail is better than +e

tgoyne commented on .github/workflows/master-push.yml:
```diff
@@ -0,0 +1,564 @@
+        - name: Get Token
+          id: token
+          run: |
+            token=$(echo $(ruby ./scripts/xcode_cloud_helper.rb --issuer-id ${{ secrets.APPLE_STORE_CONNECT_ISSUER_ID }} --key-id ${{ secrets.APPLE_STORE_CONNECT_KEY_ID }} --pk "${{ secrets.APPLE_STORE_CONNECT_API_KEY }}" get-token))
```
Is the extra `$(echo ...)` doing something complicated here? It should be just a no-op wrapper.

tgoyne commented on .github/workflows/master-push.yml:
```diff
@@ -0,0 +1,564 @@
+          with:
+            xcode-version: ${{ env.XCFRAMEWORK_VERSION }} 
+        - name: Restore ios xcframework
+          uses: actions/cache/restore@v3
```
Why is this using caches instead of artifacts to pass files between jobs? It seems like there's a lot of room for things to go very weird due to it not being at all what caches are intended for.

dianaafanador3 commented:
> I don't understand why there's now so many intermediate steps in assembling the release packages. Doing several passes of downloading artifacts, unzipping them, shuffling some files around, zipping them, and reuploading them makes it way more complicated and is probably significantly slower.

Your were right about this, I was making this a little bit too complicated, I simplified this and there is only one extra step which has to be done because the only way to set the configuration is using the name of the Xcode cloud workflow we are running.
Also, I changed from using Github's cache to artifacts as suggested.
I added a retry action to some jobs steps, because Xcode cloud API is not as stable as it should be and sometime throws unexpected 500 error.
This why it took me sometime to test this, because Xcode cloud is pretty unstable when accessed through the API.

dianaafanador3 commented:
@tgoyne Running the prepare-release workflow, takes around 1:10 minutes (https://github.com/realm/realm-swift/actions/runs/7609539064) which compared to Jenkins (https://ci.realm.io/job/cocoa-pipeline/) seems like a big improvement

dianaafanador3 commented:
Given the instability of XCode cloud, I tested generating the frameworks in Github action but in our current plan we don't support machines where we can install visionOS
